---
title: "LEA tutorial - SSMPG2025"
author: "Olivier François"
date: "July 2025"
output:
  prettydoc::html_pretty:
    self_contained: true
    theme: cayman
    highlight: github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

LEA is an R package dedicated to landscape population genomics and genotype-environment association (GEA) methods. It provides tools for analyzing population structure (snmf), including imputation of missing genotypes, performing genome-wide tests for local adaptation (lfmm2), and computing measures of genetic offset based on current or predicted environmental conditions (genetic.offset). The package is described in Frichot and François (2015) and Gain and François (2021). The  genetic offset statistics implemented in LEA are described in Gain et al. (2023).

To begin analyzing population genomic data with LEA, load the latest version of the package in R.


```{r}
# To install LEA, remove the comment symbol and run the two lines of code below
# if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
# BiocManager::install("LEA")

library(LEA)
```

This tutorial uses simulated data to illustrate the main functions of LEA, including a brief analysis of population structure, a GEA analysis of the genotypic data using four environmental variables, and the computation of genetic offset statistics for those genotypes.

 

## Example data and population struture analysis


The example dataset includes a genotypic matrix in LFMM format, containing 200 diploid individuals (rows) genotyped at 510 SNPs (columns). The corresponding environmental matrix includes 4 correlated environmental variables (env), each measured for every individual.

The data were simulated using a spatially explicit evolutionary scenario implemented in the individual-based simulation program SLiM 3 (Haller and Messer 2019). The demographic scenario involved range expansion from a single origin, with subsequent adaptation to local environmental conditions. Natural selection was simulated using a polygenic architecture, where adaptive traits were influenced by additive effects from two (unobserved) environmental variables. According to the simulation, approximately $30\%$ of the loci are affected by environmental selection.

```{r}
# loading the simulated data in R
data("offset_example")

# Y containes genotypes for 200 individuals
Y <- offset_example$geno

# X containes 4 environmental variables for 200 individuals
X <- offset_example$env
```


Population structure can be analyzed using PCA or the snmf() function, which produces results similar to those from the popular program STRUCTURE. PCA is performed below. The scree plot from the PCA suggests the presence of three main genetic groups. 

```{r}
# PCA of genotypes 
pc <- prcomp(Y)

# scree plot
plot(pc, col = "blue")
```


The scree plot of the PCA on the environmental data also suggests the presence of three main types of environmental conditions.

```{r}
# PCA of environmental variables
pc_env <- prcomp(X)

# scree plot
plot(pc_env, col = "orange")
```

The next figure shows individual samples projected onto the first principal components of the genetic matrix, with points colored according to similarity in environmental conditions.


```{r}
# clustering based on similarity in environmental conditions (3 groups)
cl <- kmeans(X, centers = 3)$cluster

# Population genetic structure colored according to similarity in environmental conditions
my_colors <- c("orange", "blue", "olivedrab")
plot(pc$x, pch = 19, col = my_colors[cl])
```


The snmf() function estimates ancestry coefficients from the genotypic matrix. The method is described in Frichot et al. (2014). The SNMF analysis is performed below.


```{r, include = FALSE}
# conversion of the genotype matrix in 'geno' format
write.geno(Y, "genotypes.geno")

# snmf runs for K = 1 to 10
proj_snmf <- snmf("genotypes.geno",
                  K = 1:10, 
                  entropy = TRUE, 
                  repetitions = 1,
                  project = "new")
```


The method suggests the presence of four (or more) genetic clusters, which is consistent with the isolation-by-distance model used in the simulated data.

```{r}
# plot the cross-entropy criterion for all runs of the project
plot(proj_snmf, cex = 1.2, col = "lightblue", pch = 19)
```

The barchart() function can be used to display a barplot of the Q-matrix, containing ancestry coefficients for each sample. 

```{r}
# display the Q-matrix

my.colors <- c("tomato", "lightblue", 
              "olivedrab", "gold")


# default is sort.by.Q = TRUE
barchart(proj_snmf, K = 4, 
        sort.by.Q = TRUE, 
        border = NA, space = 0, col = my.colors, 
        xlab = "Individuals", ylab = "Ancestry proportions", 
        main = "Ancestry matrix") -> bp

# "bp$order" contains the ordering of samples         
axis(1, at = 1:length(bp$order), 
    labels = bp$order, las = 3, cex.axis = .3)

###
```


## Genotype-Environment Association (GEA) studies with LFMM

In LEA, genotype-environment association (GEA) is performed using latent factor mixed models (LFMM)—a multivariate regression approach that estimates the effects of environmental variables while accounting for unobserved confounders known as latent factors. LFMM incorporates both fixed effects (environmental variables) and latent effects (population structure) (Frichot et al. 2013; Caye et al. 2019).

The motivation for using LFMMs is that neutral population structure is widely recognized as a major confounder when estimating the effects of natural selection driven by environmental pressures. Accurately evaluating neutral population structure from population genomic data is challenging, as it requires distinguishing between loci that are truly neutral and those under selection—an inherently difficult task.

LFMM avoids the issue of circularity by estimating the effects of environmental variables and latent factors simultaneously. The latent factors are inferred from both the genomic and environmental data and can be interpreted as the best estimates of the confounding effects due to neutral population structure. This joint modeling approach yields environmental effect size estimates with minimal bias.

The analysis of population structure suggests that $K=3$ factors is a reasonable choice for the number of latent factors in LFMM (note that K does not need to be estimated precisely).

```{r}
# use lfmm2 to estimate the latent factors from the data 
mod_lfmm2 <- lfmm2(input = Y,
                   env = X, 
                   K = 3)
```


Once the LFMM is fitted to the genomic data, genotype-environment association (GEA) significance values can be obtained using standard Student’s t-tests (full = FALSE) or Fisher’s tests (full = TRUE). The full option indicates whether all environmental variables are included simultaneously in the test.

```{r}
# GEA analysis computing locus-specific p-values, P
pv_lfmm2 = lfmm2.test(object = mod_lfmm2,
                      input = Y,
                      env = X, 
                      full = TRUE)$pvalues

# Manhattan plot showing -log P
plot(-log10(pv_lfmm2), cex = .3, xlab = "Locus",  ylab = "-Log(P)", col = "blue")
```


To decide which loci show significant association, the false discovery rate (FDR) control procedure can be applied. Here, an unusually high threshold  ($q = 20\%$) is justified because the list of candidate loci obtained after multiple testing correction (indicated by red circles) is very short. The orange line represents the Bonferroni threshold for a type I error rate of $10\%$.

```{r}
# FDR control: computing qvalues
qv_lfmm2  <- qvalue::qvalue(pv_lfmm2, fdr.level = 0.2)

# Manhattan plot showing the most interesting targets
plot(-log10(pv_lfmm2), cex = .3, xlab = "Locus",  ylab = "-Log(P)", col = "blue")

# Orange line: Bonferonni multiple testing threshold for significance
abline(h = -log10(0.1/ncol(Y)), col = "orange")

# the most interesting targets
candidates <- which(qv_lfmm2$significant)

# Circles: Candidate loci at FDR level = 30%
points(candidates, -log10(pv_lfmm2)[candidates], cex = .9, col = "brown")
```


The candidate list contains 21 loci. Since the true status of each locus is known, the result is somewhat disappointing. The list includes 3 false positives, corresponding to an observed false discovery rate (FDR) of $\frac{3}{21} \approx 14.3\%$, which is close to the targeted value ($\frac{4}{21}$). However, statistical power is very low (around 10%), indicating that most loci influenced by environmental selection were not detected in the GEA study.




## Genomic offset in LEA 


Several approaches incorporate genomic information on local adaptation into predictive measures of population maladaptation across ecological changes (Fitzpatrick et al. 2015; Waldvogel et al. 2020; Capblancq et al. 2020). These predictive measures, known as **genomic offset**, first estimate a statistical relationship between environmental gradients and allele frequencies using genotype-environment association (GEA) models. This inferred relationship is then used to evaluate differences in predicted allele frequencies between pairs of points within the ecological niche.

In LEA, we developed a **geometrical framework** for genomic offset, arguing that measures traditionally defined as distances in genetic space (i.e., differences between allele frequencies) can alternatively be viewed as distances in environmental space. By considering the duality between genetic and environmental spaces, our approach links genomic offset measures to the geometry of the ecological niche. This framework is implemented in the function `genetic.offset`.

For two environmental vectors ${\bf x}$ and ${\bf y}$ in $d$-dimensional space (here $d = 4$, the number of variables), the genetic offset is computed as follows

$$
G^2({\bf x} , {\bf y}) = ({\bf x - y}) {\bf C}_{\bf b}  ({\bf x - y})^T ,
$$

where ${\bf C}_{\bf b} = \mathbb{E}[{\bf b} {\bf b}^T]$  is the $d \times d$ covariance matrix of environmental effect sizes. Equivalently, the genetic offset can be computed as the variance of the difference between predicted allelic frequencies given environmental vectors ${\bf x}$ and ${\bf y}$

$$
G^2({\bf x} , {\bf y}) = \mathbb{V}[ p({\bf x})  -   p({\bf y}) ] ,
$$

where the variance is taken over all genetic loci considered in the analysis.  


In the simulated data, an instantaneous change in environmental conditions was applied at the end of the simulation, and the relative fitness of individuals was measured at each sampled location following this change. The modified environmental conditions can be loaded as follows.


```{r}
## modified environment
X_pred <- offset_example$env.pred
```

The genetic offset can be computed at each sampled location (as shown below) or at any location (not shown here; see the tutorial on spatial predictions).

```{r}
## computes genetic offset using all loci
g_offset <- genetic.offset(input = Y, 
                           env = X, 
                           pred.env = X_pred,
                           K = 3)
```                        

Now, we can plot the genetic offset against the Euclidean environmental distance to visualize their correlation.


```{r}
# plote genetic offset vs squared  Euclidean  environmental  distance
  Delta = X - X_pred
  dist_env2 =  rowSums(Delta^2)   
  plot(dist_env2, g_offset$offset, 
       xlab ="Squared Euclidean distance",  
       ylab ="Genetic offset", cex = .6, col = "blue")
```

The relative fitness (loss) after the environmental change can then be compared to the genetic offset. 

```{r}
## load log of relative fitness
log_relative_fitness <- read.table("fitness_variation.txt", header = FALSE)[,1]

# plot the relative fitness vs squared root of the genetic offset 
plot(g_offset$offset, - log_relative_fitness , 
     ylab ="Relative fitness (log)",  
     xlab ="Genetic offset", cex = .6, col = "blue")
```

The squared correlation between the genetic offset and the negative logarithm of relative fitness is equal to


```{r}
cor(g_offset$offset, log_relative_fitness)^2
```

The squared correlation between the  environmental Euclidean distance and the  negative log of relative fitness

```{r}
cor(dist_env2, log_relative_fitness)^2
```



## Genomic offset from candidate loci

The genetic offset analysis can be rerun with a subset of 21 candidate loci obtained from the GEA study.

```{r}
g_offset_candidate <- genetic.offset(input = Y, 
                                     env = X, 
                                     pred.env = X_pred,
                                     candidate.loci = candidates,
                                     K = 3)
                        

# plot the negative log of relative fitness vs genetic offset (candidate loci)

plot(g_offset_candidate$offset, -log_relative_fitness,
     ylab ="Relative fitness (log)",  
     xlab ="GO", cex = .6, col = "blue")
```




## Importance of variables 

Next, we can evaluate the importance of the environmental variables in the analysis. To do this, the environmental data should be scaled. For example, if temperature is measured in degrees Celsius and precipitation in millimeters, it is better to work with unitless variables. Note that scaling the environmental variables does not affect the genetic offset calculations.


```{r}
g_offset_scaled <- genetic.offset(input = Y, 
                           env = X, 
                           pred.env = X_pred,
                           scale = TRUE,
                           K = 3)
                           
# Scaling does not change genetic offsets
plot(g_offset$offset, g_offset_scaled$offset,  cex = .6)   
```

Scaling is useful for evaluating the relative importance of environmental variables. In the example studied, only two dimensions of the environmental space significantly influence the genetic offset.


```{r}                          
barplot(g_offset_scaled$eigenvalues, col = "orange", xlab = "Axes", ylab = "Eigenvalues")
```

The loadings for the first two variable combinations indicate their relative contributions to the offset. We observe that the first two variables have greater importance compared to the others.

```{r}
round(g_offset_scaled$vectors[,1:2]^2, 2)
```



## References

- Capblancq, T., Fitzpatrick, M. C., Bay, R. A., Exposito-Alonso, M.,  Keller, S. R. (2020). Genomic prediction of (mal)adaptation across current and future climatic landscapes. Annual Review of Ecology, Evolution, and Systematics, 51, 245-269. 

- Caye, K., Jumentier, B., Lepeule, J., Fran\c cois, O. (2019). LFMM 2: fast and accurate inference of gene-environment associations in genome-wide studies. Molecular Biology and Evolution, 36(4), 852-860.

- Fitzpatrick, M. C.,  Keller, S. R. (2015). Ecological genomics meets community-level modelling of biodiversity: Mapping the genomic landscape of current and future environmental adaptation. Ecology Letters, 18(1), 1-16.

- Frichot, E., Schoville, S. D., Bouchard, G.,  Fran\c cois, O. (2013). Testing for associations between loci and environmental gradients using latent factor mixed models. Molecular Biology and Evolution, 30(7), 1687-1699.

- Frichot, E., Mathieu, F., Trouillon, T., Bouchard, G., Fran\c cois, O. (2014). Fast and efficient estimation of individual ancestry coefficients. Genetics, 196(4), 973-983.

- Frichot, E.,  Fran\c cois, O. (2015). LEA: An R package for landscape and ecological association studies. Methods in Ecology and Evolution, 6(8), 925-929.
 
- Gain, C.,  Fran\c cois, O. (2021). LEA 3: Factor models in population genetics and ecological genomics with R. Molecular Ecology Resources, 21(8), 2738-2748.

- Gain, C., Rhoné, B., Cubry, P., Salazar, I., Forbes, F., Vigouroux, Y., Jay, F.,  François, O. (2023). A quantitative theory for genomic offset statistics. Molecular Biology and Evolution, 40(6), msad140.

- Haller, B., Messer, P.W. (2019). SLiM 3: Forward Genetic Simulations Beyond the Wright–Fisher Model. Molecular Biology and Evolution, 36(3), 632-637.

- Nei, M. (1973). Analysis of gene diversity in subdivided populations. Proc. Natl. Acad. Sci. USA, 70, 3321-23.
 
- Peter, B. M. (2016). Admixture, population structure, and F-statistics. Genetics, 202(4), 1485-1501.
  
- Waldvogel, A. M., Feldmeyer, B., Rolshausen, G., Exposito-Alonso, M., Rellstab, C., Kofler, R., {\it et al.} (2020). Evolutionary genomics can improve prediction of species' responses to climate change. Evolution Letters, 4(1), 4-18.